pkg/image/admission/imagepolicy/imagepolicy.go

```diff
@@ -0,0 +1,248 @@
+
+type integratedRegistryMatcher struct {
+	rules.RegistryMatcher
+}
+
+type imageResolver interface {
+	Resolve(ref imageapi.DockerImageReference) (*imageapi.Image, error)
+}
+
+// imagePolicyPlugin returns an admission controller for pods that controls what images are allowed to run on the
+// cluster.
+func newImagePolicyPlugin(client clientset.Interface, parsed *api.ImagePolicyConfig) (*imagePolicyPlugin, error) {
+	m := integratedRegistryMatcher{
+		RegistryMatcher: rules.NewRegistryMatcher(nil),
+	}
+	accepter, err := rules.NewExecutionRulesAccepter(parsed.ExecutionRules, m)
+	if err != nil {
+		return nil, err
+	}
+
+	return &imagePolicyPlugin{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+		config:  parsed,
+
+		accepter: accepter,
+
+		integratedRegistryMatcher: m,
+	}, nil
+}
+
+func (a *imagePolicyPlugin) SetDefaultRegistryFunc(fn imageapi.DefaultRegistryFunc) {
+	a.integratedRegistryMatcher.RegistryMatcher = rules.RegistryNameMatcher(fn)
+}
+
+func (a *imagePolicyPlugin) SetOpenshiftClient(c client.Interface) {
+	a.client = c
+}
+
+func (a *imagePolicyPlugin) SetProjectCache(c *cache.ProjectCache) {
+	a.projectCache = c
+}
+
+// Validate ensures that all required interfaces have been provided, or returns an error.
+func (a *imagePolicyPlugin) Validate() error {
+	if a.client == nil {
+		return fmt.Errorf("%s needs an Openshift client", api.PluginName)
+	}
+	if a.projectCache == nil {
+		return fmt.Errorf("%s needs a project cache", api.PluginName)
+	}
+	imageResolver, err := newImageResolutionCache(a.client.Images(), a.client, a.integratedRegistryMatcher)
+	if err != nil {
+		return fmt.Errorf("unable to create image policy controller: %v", err)
+	}
+	a.resolver = imageResolver
+	return nil
+}
+
+// Admit attempts to apply the image policy to the incoming resource.
+func (a *imagePolicyPlugin) Admit(attr admission.Attributes) error {
+	switch attr.GetOperation() {
+	case admission.Create, admission.Update:
+		if len(attr.GetSubresource()) > 0 {
+			return nil
+		}
+		// only create and update are tested, and only on core resources
+		// TODO: scan all resources
+		// TODO: Create a general equivalence map for admission - operation X on subresource Y is equivalent to reduced operation
+	default:
+		return nil
+	}
+
+	gr := attr.GetResource().GroupResource()
+	if !a.accepter.Covers(gr) {
+		return nil
+	}
+
+	m, err := meta.GetImageReferenceMutator(attr.GetObject())
+	if err != nil {
+		return apierrs.NewForbidden(gr, attr.GetName(), fmt.Errorf("unable to apply image policy against objects of type %T: %v", attr.GetObject(), err))
+	}
+
+	covered := make(policyAttributes)
+
+	// load exclusion rules from the namespace cache
+	var excluded sets.String
+	if ns := attr.GetNamespace(); len(ns) > 0 {
+		if ns, err := a.projectCache.GetNamespace(ns); err == nil {
+			if value := ns.Annotations[api.IgnorePolicyRulesAnnotation]; len(value) > 0 {
+				excluded = sets.NewString(strings.Split(value, ",")...)
```

Comment:
I'm betting we just make this a big switch later on to avoid coupling to it to the master-config, but the field is alpha so I don't object.